weave-js/src/components/PagePanelComponents/Home/Browse3/datasets/EditableDatasetView.tsx

```diff
@@ -0,0 +1,500 @@
+    },
+    [setDeletedRows]
+  );
+
+  const deleteRow = useCallback(
+    (absoluteIndex: number) => {
+      const rowKey = `${ADDED_ROW_ID_PREFIX}${absoluteIndex}`;
+      if (addedRows.has(rowKey)) {
+        setAddedRows(prev => {
+          const updatedMap = new Map(prev);
+          updatedMap.delete(rowKey);
+          return updatedMap;
+        });
+      } else {
+        setDeletedRows(prev => [...prev, absoluteIndex]);
+      }
+    },
+    [setDeletedRows, setAddedRows, addedRows]
+  );
+
+  const deleteAddedRow = useCallback(
+    (rowId: string) => {
+      setAddedRows(prev => {
+        const updatedMap = new Map(prev);
+        updatedMap.delete(rowId);
+        return updatedMap;
+      });
+    },
+    [setAddedRows]
+  );
+
+  const handleAddRowsClick = useCallback(() => {
+    setPaginationModel(prev => ({...prev, page: 0}));
+    setAddedRows(prev => {
+      const updatedMap = new Map(prev);
+      const newId = `${ADDED_ROW_ID_PREFIX}${uuidv4()}`;
+      const newRow = {
+        id: newId,
+        ...Object.fromEntries(initialFields.map(field => [field, ''])),
+      };
+      updatedMap.set(newId, newRow);
+      return updatedMap;
+    });
+  }, [setAddedRows, initialFields]);
+
+  const dataAsListOfDict = useMemo(() => {
+    return (loadedRows ?? []).map((row, index) => {
+      const absoluteIndex =
+        index + paginationModel.pageSize * paginationModel.page;
+      if (row == null) {
+        return {_index: absoluteIndex};
+      }
+
+      let val = row;
+      if (typeof val === 'object' && !Array.isArray(val)) {
+        if ('val' in val) {
+          val = val.val;
+        }
+        return {
+          ...flattenObjectPreservingWeaveTypes(val),
+          _index: absoluteIndex,
+        };
+      }
+      return {
+        '': val,
+        _index: absoluteIndex,
+      };
+    });
+  }, [loadedRows, paginationModel.page, paginationModel.pageSize]);
+
+  const rows = useMemo(() => {
+    if (fetchQueryLoaded) {
+      return dataAsListOfDict.map((row, i) => {
+        const digest = loadedRows[i].digest;
+        const editedRow = editedCellsMap.get(row._index);
+        return {
+          id: `${digest}_${row._index}`,
+          ...(editedRow ? {...row, ...editedRow} : row),
+        };
+      });
+    }
+    return [];
+  }, [dataAsListOfDict, loadedRows, fetchQueryLoaded, editedCellsMap]);
+
+  const combinedRows = useMemo(() => {
+    if (
+      !isEditing ||
+      numAddedRows <= paginationModel.page * paginationModel.pageSize
+    ) {
+      return rows;
+    }
+    const startIndex = paginationModel.page * paginationModel.pageSize;
+    const endIndex = startIndex + paginationModel.pageSize;
+    const displayedAddedRows = Array.from(addedRows.values())
+      .slice(startIndex, endIndex)
+      .map(row => ({...row, id: row.id}));
+    return [...displayedAddedRows, ...rows];
+  }, [rows, addedRows, numAddedRows, paginationModel, isEditing]);
+
+  const columns = useMemo(() => {
```

Comment:
not 100% sure where this is happening but the row heights aren't centered? Like the space above the trashcan isn't the same as the space beneath the trashcan. Same with the text in the rows, there is more space beneath the line than above.

<img width="795" alt="Screenshot 2025-01-09 at 10 06 26 AM" src="https://github.com/user-attachments/assets/a9f4af47-fc6b-4f9c-945e-c5ede1eac945" />

Reply:
<img width="612" alt="Screenshot 2025-01-09 at 10 09 48 AM" src="https://github.com/user-attachments/assets/d87e53f3-57be-4d79-b0bd-4b0dccb1491d" />

Reply:
I have fixed this but its still a little off for cells in the edit mode. I'm having a hard time figuring out how to style the default editing component to not do that so I may just write our own cell editing component.